src/solvers/dgsem_tree/dg_2d_parabolic.jl

```diff
@@ -0,0 +1,652 @@
+                                              equations_parabolic, surface_integral, dg, cache,
+                                              4, firsts[4], lasts[4])
+end
+function calc_divergence_boundary_flux_by_direction!(surface_flux_values::AbstractArray{<:Any,4}, t,
+                                                     boundary_condition,
+                                                     equations_parabolic::AbstractEquationsParabolic,
+                                                     surface_integral, dg::DG, cache,
+                                                     direction, first_boundary, last_boundary)
+  @unpack surface_flux = surface_integral
+
+  # Note: cache.boundaries.u contains the unsigned normal component (using "orientation", not "direction")
+  # of the viscous flux, as computed in `prolong2boundaries` of `calc_divergence!``
+  @unpack u, neighbor_ids, neighbor_sides, node_coordinates, orientations = cache.boundaries
+
+  @threaded for boundary in first_boundary:last_boundary
+    # Get neighboring element
+    neighbor = neighbor_ids[boundary]
+
+    for i in eachnode(dg)
+      # Get viscous boundary fluxes
+      flux_ll, flux_rr = get_surface_node_vars(u, equations_parabolic, dg, i, boundary)
+      if neighbor_sides[boundary] == 1 # Element is on the left, boundary on the right
+        flux_inner = flux_ll
+      else # Element is on the right, boundary on the left
+        flux_inner = flux_rr
+      end
+
+      x = get_node_coords(node_coordinates, equations_parabolic, dg, i, boundary)
+
+      # TODO: add a field in `cache.boundaries` for gradient information.
+      # Here, we pass in `u_inner = nothing` since we overwrite cache.boundaries.u with gradient information.
+      # This currently works with Dirichlet/Neuman boundary conditions for LaplaceDiffusion2D and
+      # NoSlipWall/Adiabatic boundary conditions for CompressibleNavierStokesEquations2D as of 2022-6-27.
+      # It will not work with implementations which utilize `u_inner` to impose boundary conditions.
+      flux = boundary_condition(flux_inner, nothing, get_unsigned_normal_vector_2d(direction),
+                                x, t, Divergence(), equations_parabolic)
+
+      # Copy flux to left and right element storage
+      for v in eachvariable(equations_parabolic)
+        surface_flux_values[v, i, direction, neighbor] = flux[v]
+      end
+    end
+  end
+
+  return nothing
+end
+
+function calc_gradient!(u_grad, u, t,
+                        mesh::Union{TreeMesh{2}, P4estMesh{2}}, equations_parabolic,
+                        boundary_conditions_parabolic, dg::DG, cache, cache_parabolic)
+  # Reset du
+  @trixi_timeit timer() "reset ∂u/∂t" begin
+    reset_du!(u_grad[1], dg, cache)
+    reset_du!(u_grad[2], dg, cache)
+  end
+
+  # Calculate volume integral
+  @trixi_timeit timer() "volume integral" begin
+    @unpack derivative_dhat = dg.basis
+    @threaded for element in eachelement(dg, cache)
+
```

Comment:
maybe use
```julia
gradients_x, gradients_y = u_grad
```
here and elsewhere to make the code easier to read?

Reply:
Done in da319b9f869c2e50df9b4b04b63528b9bced2eed